src/compiler/factory/emitHelpers.ts

```diff
@@ -1367,6 +1387,66 @@ export const classPrivateFieldInHelper: UnscopedEmitHelper = {
             };`
 };
 
+/**
+ * @internal
+ */
+export const addDisposableResourceHelper: UnscopedEmitHelper = {
+    name: "typescript:addDisposableResource",
+    importName: "__addDisposableResource",
+    scoped: false,
+    text: `
+        var __addDisposableResource = (this && this.__addDisposableResource) || function (env, value, async) {
+            if (value !== null && value !== void 0) {
+                if (typeof value !== "object") throw new TypeError("Object expected.");
+                var dispose;
+                if (async) {
+                    if (!Symbol.asyncDispose) throw new TypeError("Symbol.asyncDispose is not defined.");
+                    dispose = value[Symbol.asyncDispose];
+                }
+                if (dispose === void 0) {
+                    if (!Symbol.dispose) throw new TypeError("Symbol.dispose is not defined.");
+                    dispose = value[Symbol.dispose];
+                }
+                if (typeof dispose !== "function") throw new TypeError("Object not disposable.");
+                env.stack.push({ value: value, dispose: dispose, async: async });
+            }
+            else if (async) {
+                env.stack.push({ async: true });
+            }
+            return value;
+        };`
+};
+
+/**
+ * @internal
+ */
+export const disposeResourcesHelper: UnscopedEmitHelper = {
+    name: "typescript:disposeResources",
+    importName: "__disposeResources",
+    scoped: false,
+    text: `
+        var __disposeResources = (this && this.__disposeResources) || function (env) {
+            function fail(e) {
+                env.error = env.hasError ? new SuppressedError(e, env.error) : e;

```

Comment:
Good point. It will error either way, but this should probably have the same check as we do for `Symbol.dispose` in `__addDisposableResource`.

Reply:
Actually, there's a slight difference that makes the check unnecessary. In the case of `Symbol.dispose`, if we didn't check then the error you would get would just be `Object not disposable`, with no indication as to why since `Symbol.dispose` would return `undefined`, which gets coerced to the string `"undefined"` when used as a property key. In the case of `new SuppressedError`, you'll get the error `ReferenceError: SuppressedError is not defined".

We don't have a special-case error for `Promise` in our helpers either, since you'll get a useful error in those cases as well.